src/ape/managers/chain.py

```diff
@@ -389,49 +389,77 @@ class AccountHistory(BaseInterfaceModel):
     """
 
     address: AddressType
-    _transaction_list: List[ReceiptAPI] = []
+    _session_receipts: List[ReceiptAPI] = []
 
     @property
     def outgoing(self) -> Iterator[ReceiptAPI]:
         """
-        All outgoing transactions.
+        All outgoing transactions, from earliest to latest.
         """
         explorer = self.provider.network.explorer
         explorer_receipts = (
             [r for r in explorer.get_account_transactions(self.address)] if explorer else []
         )
         for receipt in explorer_receipts:
-            # TODO: Use cache
-            if receipt.txn_hash not in [t.txn_hash for t in self._transaction_list]:
+            if receipt.sender != self.address:
+                # Likely ``self.address`` is a contract.
+                # Cache the receipts by their sender instead and skip them here.
+                self.chain_manager.history.append(receipt)
+                continue
+
+            elif receipt.txn_hash not in [t.txn_hash for t in self._session_receipts]:
                 yield receipt
 
     @property
     def all(self) -> Iterator[ReceiptAPI]:
         """
-        All account transactions.
+        All account transactions, from earliest to latest.
         """
-        yield from self.outgoing
-        yield from self._transaction_list
+        sessional = sorted(self._session_receipts, key=lambda x: x.block_number)
+        for receipt in self.outgoing:
+            if not sessional or receipt.block_number <= sessional[0].block_number:
+                yield receipt
+            else:
+                yield sessional.pop(0)
 
+                # Catch up before yielding outgoing receipt.
+                if sessional:
+                    for i in range(len(sessional)):
+                        if sessional[i].block_number <= receipt.block_number:
+                            yield sessional.pop(i)
+                        else:
+                            break
 
-class TransactionHistory(BaseManager):
-    """
-    A container mapping Transaction History to the transaction from the active session.
-    """
+                # All sessional receipts before this outgoing receipt have been yielded.
+                yield receipt
 
-    _address_list_map: Dict[AddressType, AccountHistory] = {}
+        # Yield remaining sessional.
+        for sessional_receipt in sessional:
+            yield sessional_receipt
 
-    # Facilitates faster look-ups.
-    _hash_to_receipt_map: Dict[str, ReceiptAPI] = {}
+    def __iter__(self) -> Iterator[ReceiptAPI]:  # type: ignore[override]
+        yield from self.all
```

Comment:
This is better as `self.outgoing` because the idea is that `__getitem__` would be iterable by an account's `nonce` (which is the same semantically speaking)

Reply:
It would be great if Ethereum created an RPC for this. Etherscan saves the day at least, but it only works of the Etherscan plugin is installed.

I found this gem: https://github.com/ethereum/go-ethereum/issues/1897

Reply:
Nice! Long term, my idea was to use this query type:

https://github.com/ApeWorX/ape/blob/0e31f26578aa41d9015489de5ff1dd6f9ee725da/src/ape/api/query.py#L97

So, Etherscan could implement this functionality as a query plugin to source this info

But this is an implementation detail, we can refactor how this works after 0.6.0 and when the query system works a little better

Reply:
Perhaps a nice side effect of rebasing to use this though is that we can make this type of query unhandled by default unless you have a query plugin installed that handles it